lib/maths/CPoissonMeanConjugate.cc

```diff
@@ -642,6 +642,11 @@ void CPoissonMeanConjugate::sampleMarginalLikelihood(std::size_t numberSamples,
                                 m_Offset;
 
                 LOG_TRACE(<< "sample = " << sample);
+                if (sample < 0) {
+                    // We discard negative samples as they are logically incongruous for a counting distribution.
+                    // These are extremely rare but can occur when the variance is large and the offset is 0, for example.
+                    continue;
+                }
```

Comment:
I don't think this is quite right. I think `lastPartialExpectation = partialExpectation;` does need executing at the end of the loop in this case. I would propose instead changing the following lines to:
```
if (sample >= support.first && sample <= support.second) {
    samples.push_back(sample);
} else if (sample < support.first) {
    ...
    // fall through
} else {
    LOG_ERROR(<< "Sample out of bounds: sample = " << sample << ", support = ["
                           << support.first << "," << support.second << "]"
                           << ", mean = " << mean << ", variance = " << variance
                           << ", q = " << q << ", x(q) = " << xq);
}
```